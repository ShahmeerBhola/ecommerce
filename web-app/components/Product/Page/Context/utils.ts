import isEmpty from 'lodash/isEmpty';

import type { TProductContext } from './types';

export const getSingleAttributes = (_attributes?: StringMap<string[]>): StringMap => {
    let singleAttributes = {};

    for (const [key, value] of Object.entries(_attributes || {})) {
        if (value.length === 1) {
            singleAttributes = {
                ...singleAttributes,
                [key]: value[0],
            };
        }
    }

    return singleAttributes;
};

export const getAvailableAttributes = (
    variations: WordPress.ProductVariation[],
    newAttributes: StringMap,
): StringMap<string[]> => {
    const availableAttribute: StringMap<string[]> = {};

    for (const { attributes } of variations) {
        let isAvailableAttribute = true;

        for (const [key, value] of Object.entries(newAttributes)) {
            isAvailableAttribute = isAvailableAttribute && attributes[key] === value;
        }

        if (isAvailableAttribute) {
            for (const [key, value] of Object.entries(attributes)) {
                const currentAttribute = availableAttribute[key] || [];

                if (currentAttribute.indexOf(value) > -1) {
                    continue;
                }

                availableAttribute[key] = [...currentAttribute, value];
            }
        }
    }

    return availableAttribute;
};

const getAttributes = (
    variations: WordPress.ProductVariation[],
    selectedAttributes: {} | StringMap,
): TProductContext.CheckAttributesType => {
    const availableAttributes = getAvailableAttributes(variations, selectedAttributes);
    const singleAttributes = getSingleAttributes(availableAttributes);

    return {
        currentAttributes: {
            ...selectedAttributes,
            ...singleAttributes,
        },
        availableAttributes,
    };
};

export const checkAttributes = (state: TProductContext.State, payload: any): TProductContext.CheckAttributesType => {
    const { currentAttributes, product } = state;
    const { variations } = product as WordPress.Product;
    const newAttributes = {
        ...currentAttributes,
        ...payload,
    };

    const attributes = getAttributes(variations, newAttributes);

    if (isEmpty(attributes.availableAttributes)) {
        const resetAttributes = {
            ...payload,
        };

        return getAttributes(variations, resetAttributes);
    }

    return attributes;
};

export const isActiveAttribute = (attributes: StringMap, currentAttributes: StringMap): boolean => {
    return Object.entries(attributes).reduce<boolean>((isSelectedAttribute, attribute) => {
        const [key, value] = attribute;
        const matchedAttribute = value === '' || currentAttributes[key] === value;

        return isSelectedAttribute && matchedAttribute;
    }, true);
};
