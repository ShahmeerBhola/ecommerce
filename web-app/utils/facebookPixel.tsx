import React from 'react';
import config from '../config';

const { facebookPixelID } = config;

// https://www.facebook.com/business/help/402791146561655?id=1205376682832142
const event = <P,>(eventName: string, params?: P): void => {
  if (facebookPixelID !== null) {
    if (params !== undefined) {
      window.fbq('track', eventName, params);
    } else {
      window.fbq('track', eventName);
    }
  }
};

export const addLineItemToCart = ({ sku, name, price }: WordPress.ProductBase, type = 'line_item'): void =>
  event<facebook.Pixel.AddToCartParameters>('AddToCart', {
    content_name: name,
    content_type: type,
    content_ids: [sku],
    value: price,
    currency: config.defaultCurrencyCode,
  });

export const addWheelTirePackageToCart = ({ wheel, tire, addOns }: Cart.WheelTirePackage): void => {
  const addWheelTirePackageItemToCart = (product: WordPress.ProductBase): void =>
    addLineItemToCart(product, 'wheel_tire_package');

  addWheelTirePackageItemToCart(wheel);
  addWheelTirePackageItemToCart(tire);
  addOns.map((addOn) => addWheelTirePackageItemToCart(addOn));
};

export const initiateCheckout = (): void => event('InitiateCheckout');
export const completePurchase = (params: facebook.Pixel.PurchaseParameters): void =>
  event<facebook.Pixel.PurchaseParameters>('Purchase', params);

export const FacebookPixelScripts: React.FunctionComponent = () => {
  {
    if (facebookPixelID !== null) {
      return (
        <>
          <script
            dangerouslySetInnerHTML={{
              __html: `
            !function(f,b,e,v,n,t,s)
{if(f.fbq)return;n=f.fbq=function(){n.callMethod ?
              n.callMethod.apply(n, arguments) : n.queue.push(arguments)};
if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
n.queue=[];t=b.createElement(e);t.async=!0;
t.src=v;s=b.getElementsByTagName(e)[0];
s.parentNode.insertBefore(t,s)}(window,document,'script',
'https://connect.facebook.net/en_US/fbevents.js');
 fbq('init', '${facebookPixelID}');
fbq('track', 'PageView');
          `,
            }}
          />
          <noscript>
            <img
              height="1"
              width="1"
              src={`https://www.facebook.com/tr?id=${facebookPixelID}&ev=PageView
&noscript=1`}
            />
          </noscript>
        </>
      );
    }
  }
  return null;
};
