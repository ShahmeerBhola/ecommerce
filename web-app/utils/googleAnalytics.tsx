import React from 'react';
import config from '../config';
import { isBrowserSide } from './';

const { googleAnalyticsTrackingID, googleAdsConversionID } = config;

// https://developers.google.com/analytics/devguides/collection/gtagjs/pages
// https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics
export const pageview = (url: string): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('config', googleAnalyticsTrackingID, {
      page_path: url,
    });
  }
};

// https://developers.google.com/analytics/devguides/collection/gtagjs/events
const event = (action: string, eventParams: any): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('event', action, eventParams);
  }
};

const conversionEvent = (sendTo: string, eventParams?: StringMap): void =>
  event('conversion', { send_to: sendTo, ...eventParams });

export const sendOrderConfirmationConversionEvent = (orderId: string): void =>
  conversionEvent('AW-479563831/zZZcCPqr3-gBELeg1uQB', { transaction_id: orderId });

export const sendContactFormConversionEvent = (): void => conversionEvent('AW-479563831/6i-DCKyL6-gBELeg1uQB');

export const sendRequestForQuoteEvent = (): void => conversionEvent('AW-479563831/C9PsCJv7_-gBELeg1uQB');

// https://developers.google.com/analytics/devguides/collection/gtagjs/enhanced-ecommerce
export const sendEnhancedEcommerceEvent = (
  eee: GoogleAnalytics.EnhancedEcommerceEvent,
  items: GoogleAnalytics.Product[],
): void => event(eee, { items });

export const sendEnhancedEcommercePurchaseEvent = (data: GoogleAnalytics.PurchaseEvent): void =>
  event('purchase', data);

export const sendEnhancedEcommerceCheckoutOptionEvent = (data: GoogleAnalytics.CheckoutProgressEvent): void =>
  event('set_checkout_option', data);

export const enhancedEcommerceLineItemListName = 'Line Items';
export const enhancedEcommerceWheelTirePackageListName = 'Wheel Tire Package';
export const enhancedEcommerceWheelTirePackageAddOnsListName = 'Wheel Tire Package Add Ons';

export const GoogleAnalyticsAndAdsScripts: React.FunctionComponent = () => {
  {
    if (googleAnalyticsTrackingID !== null && googleAdsConversionID !== null) {
      return (
        <>
          <script async src={`https://www.googletagmanager.com/gtag/js?id=${googleAnalyticsTrackingID}`} />
          <script
            dangerouslySetInnerHTML={{
              __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${googleAnalyticsTrackingID}', {
              page_path: window.location.pathname,
            });
            gtag('config', '${googleAdsConversionID}')
          `,
            }}
          />
        </>
      );
    }
  }
  return null;
};
