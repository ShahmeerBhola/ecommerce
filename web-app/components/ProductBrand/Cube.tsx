import React from 'react';
import Link from 'next/link';
import WhiteBox from '../WhiteBox';
import { constructUrl } from '../../utils';
import isEmpty from 'lodash/isEmpty';

type Props = {
  brand: WordPress.ProductBrand;
  linkCategorySlug: string;
};

const Cube: React.FunctionComponent<Props> = ({ brand: { name, slug, image }, linkCategorySlug }) => {
  const BrandImage = (): JSX.Element => {
    const { src, alt } = image as WordPress.Image;
    return <img src={src} alt={alt} className="max-h-6" />;
  };

  const BrandBackupText = (): JSX.Element => (
    <span className="block uppercase font-bold text-xl text-black py-10 px-5">{name}</span>
  );

  return (
    <div className="flex flex-grow p-2 box-border w-full md:w-1/2 lg:w-1/3 max-w-auto md:max-w-1/2 lg:max-w-1/3">
      <WhiteBox className="flex flex-grow px-5 py-8 justify-center text-center">
        <Link href={constructUrl({ page: 'category', extra: `${linkCategorySlug}/${slug}` })}>
          <a className="flex justify-center items-center">{isEmpty(image) ? <BrandBackupText /> : <BrandImage />}</a>
        </Link>
      </WhiteBox>
    </div>
  );
};

export default Cube;
