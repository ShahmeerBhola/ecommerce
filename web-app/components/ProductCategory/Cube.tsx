import React from 'react';
import Link from 'next/link';
import Text from '../Text';
import PhotoBackground from '../PhotoBackground';
import { constructUrl } from '../../utils';
import isEmpty from 'lodash/isEmpty';

type Props = {
  category: WordPress.ProductCategory;
};

const Cube: React.FunctionComponent<Props> = ({ category: { slug, name, image } }) => {
  const wrapperClassNames = 'flex justify-end items-end h-full w-full p-5 text-right';

  const Name = (): JSX.Element => (
    <Text color="text-white" className="uppercase" size="text-4xl" weight="font-bold">
      {name}
    </Text>
  );

  const Photo = (): JSX.Element => {
    const { src } = image as WordPress.Image;
    return (
      <PhotoBackground image={src} className={wrapperClassNames}>
        <Name />
      </PhotoBackground>
    );
  };

  const ColorBackground = (): JSX.Element => (
    <div className={`${wrapperClassNames} bg-gray-80`}>
      <Name />
    </div>
  );

  return (
    <div className="h-cube w-full p-2 box-border md:w-1/2 lg:w-1/3">
      <Link href={constructUrl({ page: 'category', extra: slug })}>
        <a>{isEmpty(image) ? <ColorBackground /> : <Photo />}</a>
      </Link>
    </div>
  );
};

export default Cube;
