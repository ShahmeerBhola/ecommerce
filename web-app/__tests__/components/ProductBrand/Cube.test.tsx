import React from 'react';
import renderer from 'react-test-renderer';
import { generateMockBrand } from '../../utils';
import Cube from '../../../components/ProductBrand/Cube';

describe('ProductBrand Cube component', () => {
  test('it renders properly', () => {
    const tree = renderer.create(<Cube size="250" brand={generateMockBrand()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
