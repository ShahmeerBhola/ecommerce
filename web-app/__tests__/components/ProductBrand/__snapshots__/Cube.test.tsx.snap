// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductBrand Cube component it renders properly 1`] = `
<div
  className="h-cube p-2 box-border 250"
>
  <a
    href="/brand/test-brand-5"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <img
      alt="Test Brand"
      src="https://images.customwheeloffset.com/brand-logos/Fuel.png"
    />
  </a>
</div>
`;
